File: ProblemSolving/450DSA/Python/src/bitmanipulation/DivideTwoIntegers.py
"""
@author Anirudh Sharma

Given a two integers say a and b, find the quotient after dividing a by b without
using multiplication, division and mod operator.
"""


def divide(a, b):
    # Sign of the quotient
    sign = -1 if (a < 0) ^ (b < 0) else 1
    # Get the absolute values
    a = abs(a)
    b = abs(b)
    # Quotient
    quotient = 0
    total = 0
    for i in range(31, -1, -1):
        if total + (b << i) <= a:
            # Update the current sum
            total += (b << i)
            # Update the quotient
            quotient |= (1 << i)
    return quotient * sign


if __name__ == "__main__":
    print(divide(10, 3))
    print(divide(-43, 8))

File: ProblemSolving/450DSA/JavaScript/src/bitmanipulation/DivideTwoIntegers.js
/**
 * @author Anirudh Sharma
 * 
 * Given a two integers say a and b, find the quotient after dividing a by b without
 * using multiplication, division and mod operator.
 */
const divide = (A, B) => {
    if (A === -2147483648 && B === -1) {
        return 2147483647;
    }
    let quotient = 0, sign = 1
    if (A < 0) {
        A = -A;
        sign = -sign;
    }
    if (B < 0) {
        B = -B;
        sign = -sign;
    }
    if (A === B) {
        return sign;
    }
    for (let i = 0, val = B; A >= B; i = 0, val = B) {
        while (val > 0 && val <= A) {
            val = B << ++i;
        }
        A -= B << i - 1;
        quotient += 1 << i - 1;
    }
    return sign < 0 ? -quotient : quotient;
};

const main = () => {
    console.log(divide(10, 3));
    console.log(divide(-43, 8));
};

main();